File: frontend/src/services/reservationService.ts
import type {
  Reservation,
  CreateHoldRequest,
  ConfirmReservationRequest,
  CancelReservationRequest
} from '../types/api';

/**
 * Reservation service for booking flow
 */
export class ReservationService {
  /**
   * Create a hold on a court
   */
  async createHold(_request: CreateHoldRequest): Promise<Reservation> {
    // TODO: Implement API call
    // return apiClient.post<Reservation>('/reservations/hold', request);
    throw new Error('Not implemented yet');
  }

  /**
   * Confirm a held reservation
   */
  async confirmReservation(_reservationId: string, _request: ConfirmReservationRequest): Promise<Reservation> {
    // TODO: Implement API call
    // return apiClient.post<Reservation>(`/reservations/${reservationId}/confirm`, request);
    throw new Error('Not implemented yet');
  }

  /**
   * Cancel a reservation
   */
  async cancelReservation(_reservationId: string, _request: CancelReservationRequest): Promise<Reservation> {
    // TODO: Implement API call
    // return apiClient.post<Reservation>(`/reservations/${reservationId}/cancel`, request);
    throw new Error('Not implemented yet');
  }

  /**
   * Get user's reservations
   */
  async getMyReservations(_userId: string): Promise<Reservation[]> {
    // TODO: Implement API call
    // return apiClient.get<Reservation[]>(`/reservations/my?userId=${userId}`);
    throw new Error('Not implemented yet');
  }
}

export const reservationService = new ReservationService();

File: apps/backend/src/main/java/com/nadi/reservation/service/ReservationService.java
package com.nadi.reservation.service;

import com.nadi.reservation.model.Reservation;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import java.time.Instant;
import java.util.List;

/**
 * Reservation business logic for booking flow
 * Handles hold-confirm-cancel-expire pattern without payments
 */
@Service
public class ReservationService {
    
    /**
     * Create a temporary hold on a court
     * Race-safe implementation using MongoDB transactions
     * 
     * @param courtId court to reserve
     * @param startTime reservation start
     * @param endTime reservation end
     * @param userId user making reservation
     * @param playersCount number of players
     * @return created reservation with HELD status
     */
    public Reservation createHold(String courtId, Instant startTime, Instant endTime, 
                                String userId, int playersCount) {
        // TODO: Check for overlapping reservations using compound index
        // TODO: Calculate price using pricing rules
        // TODO: Create reservation with HELD status and holdExpiresAt
        // TODO: Use MongoDB transaction for race safety
        // TODO: Set holdExpiresAt to 15 minutes from now
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Confirm a held reservation
     * 
     * @param reservationId reservation to confirm
     * @param userId user confirming (must match reservation owner)
     * @return confirmed reservation
     */
    public Reservation confirm(String reservationId, String userId) {
        // TODO: Verify user owns the reservation
        // TODO: Check reservation is still HELD and not expired
        // TODO: Update status to CONFIRMED
        // TODO: Clear holdExpiresAt field
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Cancel a reservation
     * 
     * @param reservationId reservation to cancel
     * @param userId user cancelling (must match reservation owner)
     * @return cancelled reservation
     */
    public Reservation cancel(String reservationId, String userId) {
        // TODO: Verify user owns the reservation
        // TODO: Check reservation can be cancelled (not already expired)
        // TODO: Update status to CANCELLED
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Expire held reservations that have passed their hold time
     * Called by scheduled job every minute
     */
    @Scheduled(fixedRate = 60000) // Every minute
    public void expireHolds() {
        // TODO: Find reservations with holdExpiresAt < now and status = HELD
        // TODO: Update status to EXPIRED
        // TODO: Log expired reservations for monitoring
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Get user's reservations
     * 
     * @param userId user ID
     * @return list of user's reservations
     */
    public List<Reservation> getUserReservations(String userId) {
        // TODO: Fetch all reservations for user
        // TODO: Include court and venue information
        throw new UnsupportedOperationException("Not implemented yet");
    }
    
    /**
     * Check court availability for a time slot
     * 
     * @param courtId court ID
     * @param startTime start time
     * @param endTime end time
     * @return true if available
     */
    public boolean isCourtAvailable(String courtId, Instant startTime, Instant endTime) {
        // TODO: Check for overlapping confirmed or held reservations
        // TODO: Consider hold expiration times
        throw new UnsupportedOperationException("Not implemented yet");
    }
}
